docs/architecture/adr-070-collections-state-layer.md

```diff
@@ -0,0 +1,117 @@
+# ADR 070: Collections, a simplified storage layer for cosmos-sdk modules.
+
+## Changelog
+
+* 30/11/2022: PROPOSED
+
+## Status
+
+PROPOSED - Implemented
+
+## Abstract
+
+We propose a simplified module storage layer which leverages golang generics to allow module developers to handle module
+storage in a simple and straightforward manner, whilst offering safety, extensibility and standardisation.
+
+## Context
+
+Module developers are forced into manually implementing storage functionalities in their modules, those functionalities include
+but are not limited to:
+
+- Defining key to bytes formats.
+- Defining value to bytes formats.
+- Defining secondary indexes.
+- Defining query methods to expose outside to deal with storage.
+- Defining local methods to deal with storage writing.
+- Dealing with genesis imports and exports.
+- Writing tests for all the above.
+
+
+This brings in a lot of problems:
+- It blocks developers from focusing on the most important part: writing business logic.
+- Key to bytes formats are complex and their definition is error-prone, for example:
+  - how do I format time to bytes in such a way that bytes are sorted?
+  - how do I ensure when I don't have namespace collisions when dealing with secondary indexes?
+- The lack of standardisation makes life hard for clients, and the problem is exacerbated when it comes to providing proofs for objects present in state. Clients are forced to maintain a list of object paths to gather proofs.
+
+### Current Solution: ORM
+
+The current SDK proposed solution to this problem is [ORM](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-055-orm.md).
+Whilst ORM offers a lot of good functionality aimed at solving these specific problems, it has some downsides:
+- It requires migrations.
+- It uses the newest protobuf golang API, whilst the SDK still mainly uses gogoproto. 
+- Integrating ORM into a module would require the developer to deal with two different golang frameworks (golang protobuf + gogoproto) representing the same API objects.
+- It has a high learning curve, even for simple storage layers as it requires developers to have knowledge around protobuf options, custom cosmos-sdk storage extensions, and tooling download. Then after this they still need to learn the code-generated API.
```

Comment:
I'm not sure that collections actually has a lower learning curve. A lot of learning curve related stuff is related to docs and tutorials which I'll admit ORM is sorely lacking and should be corrected. The protobuf options are fairly small though and the tooling download could be made pretty simple, so this may be a toss up

Reply:
IMHO, collections looks like a thin abstraction based on existing storage APIs, I would say it's relatively easier to grasp than ORM.

Reply:
I guess that's fair. It's a smaller abstraction closer to the kv store and the idea of golang maps. I guess I'm seeing two dimensions of complexity here - collections has a smaller learning curve but may end up being more verbose. There's more to learn for ORM up front but actual usage will likely be more succinct.

Reply:
The largest problem I have in general with the sdk right now, not only with ORM, is people are not using protobuf correctly today. I have worked and read many code bases of users and there is always a problem in the protobuf file. Telling the user they need to rely more on proto could lead to many issues we can't clearly see today. Im not sure we can solve this, unless we go out and audit all implementations, but the sdk doesn't have unlimited resources.

Reply:
This is so much simpler than the proto-ORM was

Reply:
I agree. This seems so much simpler!